ipq806x: use generic dt-based board-detection and improve image filenames

luaraneda opened the pull request:
Use the generic board detection method:
- Board name: First compatible string from the device tree
- Board model: Model property from the device tree

Change occurrences of board name in userspace by the compatible
string, and removed target specific board detection script

Replace the definition of SUPPORTED_DEVICES in Device/Default
to extract the dt compatible string from each device definition.
Additionally, for devices supported by lede-17.01, append
the value of BOARD_NAME to SUPPORTED_DEVICES in the device
definition.
Also, use <manufacturer>_<modelname> as image name.

The last commit it just a cosmetic change to sort boardname alphabetically, it
can be dropped if necessary.

- compile-tested: All ipq806x boards. Compared sysupgrade metadata
to ensure update is possible after the changes
- run-tested: Out of tree Asus AC58U

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
ipq806x: add missing dt compatible string to ap-dk04 board

Compatible string is mandatory to dts files

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
ipq806x: add a helper variable with the vendor name stripped

This will allow to maintain the current syntax for LEDs config
when switching to a device tree compatible string boardname.

None of the current boards use a comma in the boardname, so they
will be unaffected.

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
ipq806x: sync image filename with dt compatible string

Use <manufacturer>_<modelname> as image name for board using the
devicetree compat string as boardname.

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

mkresin commented:
> ipq806x: use SUPPORTED_DEVICES instead of BOARD_NAME 

You need to keep the BOARD_NAME variable, otherwise an upgrade from lede-17.01 to master will not work. The subdirectory in the tar archive is named based on this variable and lede-17.01 will look for the kernel/rootfs image in sysupgrade-<boardname>.

mkresin commented on target/linux/ipq806x/base-files/etc/board.d/01_leds:
```diff
@@ -8,39 +8,40 @@
-	ucidef_set_led_switch "wan" "wan" "${board}:white:wan" "switch0" "0x20"
-	ucidef_set_led_switch "lan" "lan" "${board}:white:lan" "switch0" "0x1e"
+avm,fritzbox-4040)
+	ucidef_set_led_wlan "wlan" "WLAN" "${boardname}:green:wlan" "phy0tpt" "phy1tpt"
```
You need to use a fixed boardname here. The LEDs are named `fritz4040:<colour>:<function>`.

luaraneda replied:
You're right. I already fixed it on another branch, but I don't wanna push the changes yet until I have addressed the BOARD_NAME variable problem.

I left the LEDs names hard-coded as you suggested (fritz4040:white:wan).
I imagine we can't change the device-tree because we would break the LEDs of existing users which already have the LEDs defined in the file /etc/config/system. Is that correct?

mkresin replied:
Changing LED names is always problematic as it requires at least a script which migrates existing configurations. Since the fritzbox 4040 doesn't have yet a stable image, it might be possible to either change the compatible string or the LED names.

Lets see what @chunkeey thinks about it, since the fritzbox 4040 commit was from him.

luaraneda commented:
@mkresin about the BOARD_NAME variable, when I created the changes, I looked only at the fwtool_check_image() function, which checks the image metadata, and works correctly when the BOARD_NAME matches one of the image's supported devices.

Is the behavior prior to commit c90a8cb7552cc7e843c328283268d67709d26a8c the same as 17.01?

I could keep the BOARD_NAME variable, and append the compatible string to SUPPORTED_DEVICES on every device, which will result in both variables being present for every device.
Is that a reasonable solution?

I just found a solution, not tested yet, but it might work:
- Define BOARD_NAME in `Device/Default` as
`BOARD_NAME = $$(firstword $$(SUPPORTED_DEVICES))`
- On every device definition, the first item of SUPPORTED_DEVICES must be the value 
that BOARD_NAME had before

That way, we would only keep the variable SUPPORTED_DEVICES on every device definition.

mkresin commented:
> Is the behavior prior to commit c90a8cb the same as 17.01?

Yes it is. c90a8cb only lifted the requirement of enforcing a fixed subdirectory name. Everything else should be the same as with 17.01.

> I could keep the BOARD_NAME variable, and append the compatible string to SUPPORTED_DEVICES on every device, which will result in both variables being present for every device.
Is that a reasonable solution?

Yes please do it this way!

> I just found a solution, not tested yet, but it might work:
>- Define BOARD_NAME in `Device/Default` as
`BOARD_NAME = $$(firstword $$(SUPPORTED_DEVICES))`
>- On every device definition, the first item of SUPPORTED_DEVICES must be the value 
that BOARD_NAME had before
>
>That way, we would only keep the variable SUPPORTED_DEVICES on every device definition.

Setting the BOARD_NAME depending on a particular order of the SUPPORTED_DEVICES looks really error prone to me and might be hard to get for people which aren't that familiar with the image build code/Makefiles. I favour easier to get/read code over saving a few (duplicate) lines.

mkresin commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -29,7 +29,6 @@ define Device/Default
 	KERNEL_INITRAMFS_PREFIX := $$(IMG_PREFIX)-$(1)-initramfs
 	KERNEL_PREFIX := $$(IMAGE_PREFIX)
 	KERNEL_LOADADDR = 0x42208000
-	SUPPORTED_DEVICES = $$(BOARD_NAME)
```
Please use `SUPPORTED_DEVICES := $(subst _,$(comma),$(1))` here. It will use the part after the `define Device/`as "default" and replaces underline by comma to match the userspace boardname. It should match the userspace boardname afterwards.

For devices which are supported by lede-17.01, like the TP-Link Archer C2600, append the former used boardname in the device define:

```
define Device/tplink_c2600
	...
	SUPPORTED_DEVICES += c2600
	...
endef
```

I'm fine to kind of (soft) break compatibility for Devices which are only supported by master. They will be upgradeable by using `sysupgrade --force /path/to/sysupgrade.bin`.

luaraneda replied:
Ok, will do.

luaraneda commented:
@mkresin I've pushed the requested changes:
- Dropped the commit _ipq806x: use SUPPORTED_DEVICES instead of BOARD_NAME_ to keep the BOARD_NAME variable.
- Replaced the boardame variable for a fixed value (fritz4040) for the fritzbox-4040 board.
- Replaced the definition of SUPPORTED_DEVICES to extract the dt compatible string from the device definition.

With the changes applied, the folder inside the .tar files are now correct, and image metadata is ok too.

I found just one oddity that didn't happen with my original changes:
The AP148 board has 2 devices defined, "normal" and legacy.
The folder inside the .tar file is correct. However, by doing the
`SUPPORTED_DEVICES := $(subst _,$(comma),$(1))` the metadata
generated for the "normal" version is 
> "supported_devices":["qcom,ipq8064-ap148","ap148"]

and for the legacy version is:
> "supported_devices":["qcom,ipq8064-ap148-legacy","ap148"]

Which I think is incorrect, because the dt compatible string for both devices is
> "qcom,ipq8064-ap148"

The problem can be solved (not implemented on these changes) by appending
the dt compatible string to the SUPPORTED_DEVICES variable of the legacy device,
but the metadata will contain one useless value (legacy):
> "supported_devices":["qcom,ipq8064-ap148-legacy","ap148","qcom,ipq8064-ap148"]

How do we proceed on this case?

mkresin commented:
>Which I think is incorrect, because the dt compatible string for both devices is
>> "qcom,ipq8064-ap148"
>
>The problem can be solved (not implemented on these changes) by appending
>the dt compatible string to the SUPPORTED_DEVICES variable of the legacy device,
>but the metadata will contain one useless value (legacy):
>> "supported_devices":["qcom,ipq8064-ap148-legacy","ap148","qcom,ipq8064-ap148"]
>
>How do we proceed on this case?

Two completely different images/kernels for the same board is a special case. Overriding the supported devices completely for the legacy image is the way to go:

```
define Device/qcom_ipq8064-ap148-legacy
	$(call Device/LegacyImage)
	$(call Device/UbiFit)
	DEVICE_DTS := qcom-ipq8064-ap148
	BLOCKSIZE := 128k
	PAGESIZE := 2048
	BOARD_NAME := ap148
	SUPPORTED_DEVICES := qcom,ipq8064-ap148 ap148
	DEVICE_TITLE := Qualcomm AP148 (legacy)
	DEVICE_PACKAGES := ath10k-firmware-qca99x0
endef
TARGET_DEVICES += qcom_ipq8064-ap148-legacy
```

luaraneda committed:
ipq806x: convert to dt-based board-detection

Use the generic board detection method:
- Board name: First compatible string from the device tree
- Board model: Model property from the device tree

Change occurrences of board name in userspace by the compatible
string, and removed target specific board detection script

Replace the definition of SUPPORTED_DEVICES in Device/Default
to extract the dt compatible string from each device definition.
Additionally, for devices supported by lede-17.01, append
the value of BOARD_NAME to SUPPORTED_DEVICES in the device
definition.

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda committed:
ipq806x: Sort occurrences of boardame alphabetically

This restores the alphabetical sort that was present
before the renaming of boardname.

Signed-off-by: Luis Araneda <luaraneda@gmail.com>

luaraneda commented:
@mkresin change done. That was my second alternative, but I was trying to avoid overriding the variable.

I think I've addressed all the requested changes with these commits

mkresin commented:
merged into the ipq806x branch of [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary),

mkresin commented:
You are correct. A downgrade is only possible using the force parameter.  It is by design not considered as bug. We only take care that an upgrade from previous version is possible.

It is similar to what we do with /etc/config/*. During upgrade the config gets updated for some targets/boards, but we never rollback any changes on downgrade. Means, a config created by a master/snapshot image might be incompatible with a 17.01 image and can cause issues if kept.

Nevertheless we took care that a downgrade using the force parameter doesn't break anything. At least for now.

After switching to kernel 4.14 you will encounter even more issues. ubi volumes created by kernel 4.14 can only be mounted as read only by kernel 4.9. Which is especially an issue for the rootfsdata volume which is (re)created during sysupgrade.
